lib/compiler-singlepass/src/compiler.rs

```diff
@@ -106,29 +106,28 @@ impl Compiler for SinglepassCompiler {
             .map(|(i, input)| {
                 let reader = wasmer_compiler::FunctionReader::new(input.module_offset, input.data);
 
-                let mut local_reader = reader.get_locals_reader()?;
-                // This local list excludes arguments.
-                let mut locals = vec![];
-                let num_locals = local_reader.get_count();
-                for _ in 0..num_locals {
-                    let (count, ty) = local_reader.read()?;
-                    for _ in 0..count {
-                        locals.push(ty);
-                    }
-                }
-
                 let mut generator = FuncGen::new(
                     module,
                     module_translation,
                     &self.config,
                     &vmoffsets,
                     &table_styles,
                     i,
-                    &locals,
                     calling_convention,
                 )
                 .map_err(to_compile_error)?;
 
+                let mut local_reader = reader.get_locals_reader()?;
+                for _ in 0..local_reader.get_count() {
+                    let (count, ty) = local_reader.read()?;
+                    // Overflows feeding a local here have most likely already been caught by the
+                    // validator, but it is possible that the validator hasn't been run at all, or
+                    // that the validator does not impose any limits on the number of locals.
```

Comment:
I am still a bit confused by the wording here. If we allow running the compiler without running the validator at all, then `expect` in the `local_type` seems wrong? 

Though, this is minor -- I think the actual logic here is OK -- we want to `.unwrap` on spec-things and `?` on limits-things. 

Ideally, we emit a specific `HardLimitExceeded` error here, so that we can `tracing::error!` it on the near side.

Reply:
> If we allow running the compiler without running the validator at all, then expect in the local_type seems wrong?

It is a programming error to feed invalid webassembly to the compiler without being sure it is valid, just like it is a programming error to index into an array at a wrong index – hence a panic. We could return an error for this class of mistakes too, but this seems like a prime example for panic use to me.

I just noticed that the number of locals is in fact bounded to u32::max_value() as well because of this wording:

> Vectors are bounded sequences of the form A<sup>n</sup> (or A<sup>∗</sup>), where the A can either be values or complex constructions. A vector can have at most 2<sup>32</sup>−1 elements. 

And 

> ​[func](https://www.w3.org/TR/wasm-core-1/#syntax-func)​::=​{[type](https://www.w3.org/TR/wasm-core-1/#syntax-func) [typeidx](https://www.w3.org/TR/wasm-core-1/#syntax-typeidx),[locals](https://www.w3.org/TR/wasm-core-1/#syntax-func) [vec](https://www.w3.org/TR/wasm-core-1/#syntax-vec)([valtype](https://www.w3.org/TR/wasm-core-1/#syntax-valtype)),[body](https://www.w3.org/TR/wasm-core-1/#syntax-func) [expr](https://www.w3.org/TR/wasm-core-1/#syntax-expr)}​​

so this overflow is also an indication of a malformed module…